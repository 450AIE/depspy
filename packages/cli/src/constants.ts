export const CONFIG_FILE = "dep-spy.config.js";
