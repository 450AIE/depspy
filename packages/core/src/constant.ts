export interface Node {
  name: string;
  version: string;
  description?: string;
  cache?: string;
  circlePath?: string;
  children: Record<string, Node>;
}
export interface Config {
  depth?: number;
  outDir?: string;
  online?: boolean;
}
