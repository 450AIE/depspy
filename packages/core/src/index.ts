export { generateGraph } from "./generateGraph";
export type { Node, Config } from "./constant";
