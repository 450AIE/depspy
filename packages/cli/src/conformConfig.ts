import path from "path";
import fs from "fs";
import { pathToFileURL } from "url";
import { CONFIG_FILE } from "./constants";
async function getLocalConfig() {
  const resolvePath = path.join(process.cwd(), CONFIG_FILE);
  if (fs.existsSync(resolvePath)) {
    const localConfig = (await import(pathToFileURL(CONFIG_FILE).toString()))
      .default;
    return localConfig;
  } else {
    return {};
  }
}

export async function conformConfig(options) {
  filterDuplicateOptions(options);
  const localConfig = await getLocalConfig();
  return { ...localConfig, ...options };
}

export const filterDuplicateOptions = <T extends object>(options: T) => {
  for (const [key, value] of Object.entries(options)) {
    if (Array.isArray(value)) {
      options[key as keyof T] = value[value.length - 1];
    }
  }
};
