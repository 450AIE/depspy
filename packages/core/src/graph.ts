import { getModuleInfo } from "@dep-spy/utils";
import { Node, Config } from "./constant";
export class Graph {
  private cache: Map<string, Node> = new Map();
  private path: Set<string> = new Set();
  constructor(
    private readonly info: string,
    private readonly config: Config = { depth: Infinity, online: false },
  ) {}
  async initGraph(info: string, father: string) {
    const { name, version, dependencies, description } = await getModuleInfo(
      info,
      father,
      this.config.online,
    );
    const id = name + "!" + version;
    //直接返回缓存
    if (this.cache.has(id)) {
      this.cache.get(id).cache = id;
      return this.cache.get(id)!;
    }
    //没有子依赖直接返回
    if (!dependencies) {
      return new GraphNode(name, version, {}, { description });
    }
    //循环依赖
    if (this.path.has(id)) {
      const circlePath = Array.from(this.path.keys()).join("/");
      return new GraphNode(name, version, {}, { description, circlePath });
    }
    //生成父节点
    const children: Record<string, Node> = {};
    const curNode = new GraphNode(name, version, children, { description });
    const dependenceNames = Object.keys(dependencies);
    //A-加入当前路径
    this.path.add(id);
    //递归生成子节点
    for (let i = 0; i < dependenceNames.length; i++) {
      if (this.config.depth && this.path.size == this.config.depth) {
        break;
      }
      const child = await this.initGraph(dependenceNames[i], name);
      const childId = child.name + "!" + child.version;
      this.cache.set(childId, child!);
      children[childId] = child;
    }
    //A-删除当前路径
    this.path.delete(id);
    return curNode;
  }
  async output() {
    return await this.initGraph(this.info, "");
  }
}

class GraphNode implements Node {
  constructor(
    public name: string,
    public version: string,
    public children: Record<string, Node>,
    otherFields: { description?: string; circlePath?: string },
  ) {
    Object.entries(otherFields).forEach(([key, value]) => {
      this[key] = value;
    });
  }
}
