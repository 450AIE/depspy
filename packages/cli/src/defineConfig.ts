import { Config } from "@dep-spy/core";
export function defineConfig(config: Config) {
  return config;
}
