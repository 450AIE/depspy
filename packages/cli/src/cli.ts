import cac from "cac";
import ora from "ora";
import { blue, green, yellow } from "chalk";
import { generateGraph } from "@dep-spy/core";
import { conformConfig } from "./conformConfig";
const cli = cac();
cli
  .command("[analysis,ana]", "解析本地项目依赖关系图")
  .option("--depth <depth>", "依赖图最大深度", {
    type: ["number"],
  })
  .option("--outDir <outDir>", "输出依赖图路径", {
    type: ["string"],
  })
  .option("--online [online]", "是否在线分析", {
    type: ["boolean"],
  })
  .action(async (_, options) => {
    const spinner = ora(blue(" 🕵️  <<<正在潜入🚀>>>")).start();
    const startTime = Date.now();
    options = await conformConfig(options);
    generateGraph("", options).then(() => {
      spinner.stop();
      console.log(green(`破解完成,耗时 ${yellow(Date.now() - startTime)} ms`));
    });
  });

cli.help();
cli.parse();
