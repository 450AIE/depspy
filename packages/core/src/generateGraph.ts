import { Node, Config } from "./constant";
import { Graph } from "./graph";
interface generateGraphRes {
  root: Node;
  codependency: Node[];
  circularDependency: Node[];
}
export async function generateGraph(): Promise<generateGraphRes>;
export async function generateGraph(info: string): Promise<generateGraphRes>;
export async function generateGraph(config: Config): Promise<generateGraphRes>;
export async function generateGraph(
  info: string,
  config: Config,
): Promise<generateGraphRes>;
export async function generateGraph(
  info?: string | Config,
  config: Config = {},
): Promise<generateGraphRes> {
  let graph: Graph | null = null;
  //实现各种重载
  if (!info) {
    graph = new Graph("", config);
  } else if (typeof info == "object") {
    graph = new Graph("", info);
  } else if (typeof info == "string") {
    graph = new Graph(info, config);
  } else {
    throw new Error(`Invalid parameters ${info}-${config}`);
  }
  const root = await graph.getGraph();
  const codependency = await graph.getCodependency();
  const circularDependency = await graph.getCircularDependency();
  await graph.outputToFile();
  return { root, codependency, circularDependency };
}
