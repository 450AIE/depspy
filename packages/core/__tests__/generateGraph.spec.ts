import { describe, expect, test } from "vitest";
import { generateGraph } from "../src";
describe("输入一个模块info和配置,输出该模块完整依赖图", () => {
  test("主函数", async () => {
    expect(await generateGraph("")).toMatchSnapshot();
  });
});
