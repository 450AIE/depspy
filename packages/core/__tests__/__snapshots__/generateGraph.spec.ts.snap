// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`输入一个模块info和配置,输出该模块完整依赖图 > 主函数 1`] = `
GraphNode {
  "children": {},
  "description": undefined,
  "name": "dep-spy",
  "version": "1.1.0",
}
`;
